docs/Blog-Post.md

```diff
@@ -0,0 +1,578 @@
+[tutorial](https://colab.research.google.com/drive/1TiTcQQ4V5mopbgCu1SVl-oqJtXn7rFnC?usp=sharing)
+for an introduction to Haliax's neural network modules, which also introduces the relevant parts of Equinox.)
+
+Haliax is modeled on Alexander Rush's [Tensor Considered Harmful](https://nlp.seas.harvard.edu/NamedTensor),
+which argues that named tensors are a better abstraction than the position-based tensors that are common in deep learning.
+In particular, he argues that:
+
+* Named axes are more semantically meaningful and rely less on bitrot-prone comments.
+* Named axes allow you to abstract over unreferenced dimensions, making code more flexible.
+* Broadcasting leads to unreadable strings of `reshape`s, `transpose`s, `view`s and (`un`)`squeeze`s that obfuscate the intent of the code.
+
+We also contend that positional code is more bug-prone: those `transpose`s and `reshape`s are easy to get wrong,
+resulting in silent bugs that are hard to catch. Implicit broadcasting, also common in positional code, can lead to
+similar easy-to-miss bugs.
+
+## A Simple, but Incorrect, Example with Positional Axes
+
+As an example, let's consider the following simple linear model. Before reading on, try to spot the bug:
+
+```python
+import jax.numpy as jnp
+import jax.random as jrandom
+from jax.random import PRNGKey
+
+# predict y from x using  a linear model (W)
+# (In JAX, you pass in the PRNGKey to get reproducible random numbers)
+x = jrandom.uniform(PRNGKey(0), (128, 64))
+y = jrandom.uniform(PRNGKey(1), (128,))
+W = jrandom.uniform(PRNGKey(2), (64, 1))
+
+def mse(pred, target):
+    return jnp.mean((pred - target) * (pred - target))
+
+y_pred = x @ W
+mse(y_pred, y)
+```
+
+This code appears straightforward, but it's incorrect: the dimensions of `y_pred` and `y` are not the same.
+Because `y_pred` is a 2D array of shape `(128, 1)`, and `y` is a 1D array of shape `(128,)`, the `-` operator will broadcast `y` to shape `(128, 128)`.
+(This makes the subtraction an "outer product"-like operation rather than the intended elementwise subtraction.)
+But, you won't get an error at runtime; this is a silent bug. The `mean` call hides the bug by averaging over all values.
+
+This is a common bug in deep learning code, and it's easy to miss. We, and anecdotally many others we have spoken to, have
+individually lost multiple days to this exact bug, and it shows up in every deep learning framework.
+
+But if we use named tensors, we avoid this bug without having to think about it. Here's what this code looks like in Haliax:
+
+```python
+import haliax as hax
+from jax.random import PRNGKey
+
+Batch = hax.Axis("batch", 128)
+Feature = hax.Axis("feature", 64)
+
+x = hax.random.uniform(PRNGKey(0), (Batch, Feature))
+y = hax.random.uniform(PRNGKey(1), Batch)
+
+def mse(pred, target):
+    return hax.mean((pred - target) * (pred - target), axis=Batch)
+
+W = hax.random.uniform(PRNGKey(2), (Feature,))
```

Comment:
Can we put W next to x, y just like in the pytorch code?  In general, make the code as isomorphic as possible.

Reply:
oops, yeah